### PR TITLE
[CCR] FollowingEngine should fail with 403 if operation has no seqno assigned

### PR DESCRIPTION
**Pull request by martijnvg**
Fail with a 403 when indexing a document directly into a follower index.

In order to test this change, I had to disable assertions in the ccr
engine package. I think that is the right trade off.

### CODE REVIEW

**Comment by dnhatn** [comment]
@martijnvg Can we minimize the scope that we will disable the assertion? I think we can create a class that does only the PreFlight then disable the assertion only that class. WDYT?

**Comment by martijnvg** [comment]
@dnhatn I've updated this PR.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineAssertions.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    static void preFlight(final Engine.Operation operation) {
+        /*
+         * We assert here so that this goes uncaught in unit tests and fails nodes in standalone tests (we want a harsh failure so that we
+         * do not have a situation where a shard fails and is recovered elsewhere and a test subsequently passes). We throw an exception so
+         * that we also prevent issues in production code.
+         */
+        assert operation.seqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO;
+        if (operation.seqNo() == SequenceNumbers.UNASSIGNED_SEQ_NO) {
+            throw new ElasticsearchStatusException("a following engine does not accept operations without an assigned sequence number",
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineAssertions.java]: I think we should keep this `check/throw` in the FollowingEngine.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineAssertions.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * Moved preflight and assertPrimaryIncomingSequenceNumber check to its own class,
+ * so that when testing writing directly into a follower index,
+ * only these assertions here need to be disabled instead of all assertions in FollowingEngine class.
+ */
+final class FollowingEngineAssertions {
+
+    static void preFlight(final Engine.Operation operation) {
+        /*
+         * We assert here so that this goes uncaught in unit tests and fails nodes in standalone tests (we want a harsh failure so that we
+         * do not have a situation where a shard fails and is recovered elsewhere and a test subsequently passes). We throw an exception so
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineAssertions.java]: Can we make this method return true (and add prefix assert) and move this comment to the FollowingEngine before we call `assert FollowingEngineAssertions. preFlight()` in the FollowingEngine?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineAssertions.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        assert operation.seqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO;
+        if (operation.seqNo() == SequenceNumbers.UNASSIGNED_SEQ_NO) {
+            throw new ElasticsearchStatusException("a following engine does not accept operations without an assigned sequence number",
+                RestStatus.FORBIDDEN);
+        }
+        assert (operation.origin() == Engine.Operation.Origin.PRIMARY) == (operation.versionType() == VersionType.EXTERNAL) :
+            "invalid version_type in a following engine; version_type=" + operation.versionType() + "origin=" + operation.origin();
+    }
+
+    static void assertPrimaryIncomingSequenceNumber(final Engine.Operation.Origin origin, final long seqNo) {
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineAssertions.java]: Can we make this method return true and call `assert assertPrimaryIncomingSequenceNumber` in the FollowingEngine.

**Review by dnhatn — COMMENTED** [review]
Thanks @martijnvg. I left some minor comments.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks @martijnvg.

**Pull request merged**